python/ray/ml/trainer.py

```diff
@@ -171,7 +188,32 @@ def preprocess_datasets(self) -> None:
         The transformed datasets will be set back in the ``self.datasets`` attribute
         of the Trainer to be used when overriding ``training_loop``.
         """
-        raise NotImplementedError
+        # Evaluate all datasets.
+        self.datasets = {k: d() if callable(d) else d for k, d in self.datasets.items()}
+
+        # Transform all datasets concurrently in remote tasks.
+        transform_task_dict = {}
+
+        transform_task = ray.remote(
+            lambda preprocessor, dataset: preprocessor.transform(dataset)
```

Comment:
This is actually problematic for fault tolerance, since then the owner of the return dataset ends up being a remote worker instead of the local trainer. This means if those nodes fail the dataset will be lost.

How about we execute the transforms locally for now and add a TODO to revisit this?

Reply:
oh interesting...thanks for letting me know.

Ok will make the change to execute locally. Thanks!

Reply:
Yep. Btw, another way is to use a thread-pool locally. But, probably safer to leave that for later cc @stephanie-wang

Reply:
Updated!